fontforge/utanvec.c

```diff
@@ -0,0 +1,237 @@
+    if ( BPNEAR(ut1, ut2) )
+	return true;
+
+    if ( BPNEAR(ut2, ut3) || BPNEAR(ut1, ut3) )
+	return false;
+
+    if (ccw) {
+	tmp = ut1;
+	ut1 = ut3;
+	ut3 = tmp;
+    }
+
+    if ( UTanVecGreater(ut1, ut3) ) {
+	return UTanVecGreater(ut1, ut2) && UTanVecGreater(ut2, ut3);
+    } else {
+	return    (UTanVecGreater(ut1, ut2) && UTanVecGreater(ut3, ut2))
+	       || (UTanVecGreater(ut2, ut1) && UTanVecGreater(ut2, ut3));
+    }
+}
+
+int JointBendsCW(BasePoint ut_ref, BasePoint ut_vec) {
+    // magnitude of cross product
+    bigreal r = ut_ref.x*ut_vec.y - ut_ref.y*ut_vec.x;
+    if ( RealWithin(r, 0.0, UTMARGIN) )
+	return false;
+    return r > 0;
+}
+
+BasePoint SplineUTanVecAt(Spline *s, bigreal t) {
+    BasePoint raw;
+
+    if ( SplineIsLinear(s) ) {
+	raw.x = s->to->me.x - s->from->me.x;
+	raw.y = s->to->me.y - s->from->me.y;
+    } else {
+	raw = SPLINEPTANVAL(s, t);
+	// For missing slopes take a very small step away and try again.
+	if ( raw.x==0 && raw.y==0 )
+	    raw = SPLINEPTANVAL(s, (t+UTRETRY <= 1.0) ? t+UTRETRY : t-UTRETRY);
+    }
+    return MakeUTanVec(raw.x, raw.y);
+}
+
+/* Return the lowest t value greater than min_t such that the
+ * tangent at the point is parallel to ut, or -1 if the tangent
+ * never has that slope.
+ */
+#define ROTY(p, ut) (-(ut).y*(p).x + (ut).x*(p).y)
+#define ZCHECK(v) (RealWithin((v),0,1e-10) ? 0 : (v))
+
+bigreal SplineSolveForUTanVec(Spline *spl, BasePoint ut, bigreal min_t) {
+    bigreal tmp, yto;
+    extended te1, te2;
+    Spline1D ys1d;
+
+    // Nothing to "solve" for; lines should be handled by different means
+    // because you can't "advance" along them based on slope changes
+    if ( SplineIsLinear(spl) )
+        return -1;
+
+    // Only check t==0 if min_t is negative
+    if ( min_t<0 && BPNEAR(ut, SplineUTanVecAt(spl, 0.0)) )
+       return 0.0;
+
+    // This is probably over-optimized
+    ys1d.d = ROTY(spl->from->me, ut);
+    yto = ROTY(spl->to->me, ut);
+    if ( spl->order2 ) {
+        ys1d.c = ZCHECK(2*(ROTY(spl->from->nextcp, ut)-ys1d.d));
+        ys1d.b = ZCHECK(yto-ys1d.d-ys1d.b);
+        ys1d.a = 0;
+    } else {
+	tmp = ROTY(spl->from->nextcp, ut);
+	ys1d.c = ZCHECK(3*(tmp-ys1d.d));
+	ys1d.b = ZCHECK(3*(ROTY(spl->to->prevcp, ut)-tmp)-ys1d.c);
+	ys1d.a = ZCHECK(yto-ys1d.d-ys1d.c-ys1d.b);
+    }
+    if ( ys1d.a!=0 && (   Within16RoundingErrors(ys1d.a+ys1d.d,ys1d.d)
+                       || Within16RoundingErrors(ys1d.a+yto,yto)))
+	ys1d.a = 0;
+
+    // After rotating by theta the desired angle will be at a y extrema
+    SplineFindExtrema(&ys1d, &te1, &te2);
+
+    if (   te1 != -1 && te1 > (min_t+1e-9)
+        && BPNEAR(ut, SplineUTanVecAt(spl, te1)))
+	return te1;
+    if (   te2 != -1 && te2 > (min_t+1e-9)
+        && BPNEAR(ut, SplineUTanVecAt(spl, te2)))
+	return te2;
+
+    // Check t==1
+    if ( (min_t+UTRETRY) < 1 && BPNEAR(ut, SplineUTanVecAt(spl, 1.0)) )
+        return 1.0;
+
+    // Nothing found
+    return -1;
+}
+
+void UTanVecTests() {
```

Comment:
Can we either move this into a test binary and call it as part of the tests, or just comment this out? With how fontforge is currently structured, even if this gets unused, it won't be dropped from the dll, since we export everything

Reply:
I don't have any strong objections to doing so, and this isn't even currently hooked up to a regression test. 

On the other hand, suppose there were a bunch of similar functions spread out through other files. What would be the concern about them? Confusion on the part of people using `nm` and the like to look at the library? (Surely not storage costs?)

Reply:
It's not a major concern tbh, but it's more of a 'repo hygine' thing, if that makes sense. I'm not worried about anyone trying to link against this, but it's kind of crap having test code exported in a releasable dll.

If I had my way and a lot more time on my hands, I'd delete all unused code and explicitly mark all exportable functions. But that's a story for another time...